Testing for Content Security Policy (#708)
OWASP/wstg: document/4-Web_Application_Security_Testing/02-Configuration_and_Deployment_Management_Testing/12-Test_for_Content_Security_Policy.md

# Testing for Content Security Policy

|ID          |
|------------|
|WSTG-CONF-12|

## Summary

> Content Security Policy is a declarative allow-list policy enforced through `Content-Security-Policy` header or equivalent `<meta>` element. It allows developers to restrict the sources from which resources such as JavaScript, CSS, images, files etc. are loaded. CSP is an effective defense in depth technique to mitigate the risk of vulnerabilities such as Cross Site Scripting (XSS) and Clickjacking.

Reviewer: ```suggestion
Content Security Policy is a declarative allow-list policy enforced through `Content-Security-Policy` response header or equivalent `<meta>` element. It allows developers to restrict the sources from which resources such as JavaScript, CSS, images, files etc. are loaded. CSP is an effective defense in depth technique to mitigate the risk of vulnerabilities such as Cross Site Scripting (XSS) and Clickjacking.
```

> Content Security Policy is a declarative allow-list policy enforced through `Content-Security-Policy` response header or equivalent `<meta>` element. It allows developers to restrict the sources from which resources such as JavaScript, CSS, images, files etc. are loaded. CSP is an effective defense in depth technique to mitigate the risk of vulnerabilities such as Cross Site Scripting (XSS) and Clickjacking.

Reviewer 2: ```suggestion
Content Security Policy (CSP) is a declarative allow-list policy enforced through `Content-Security-Policy` response header or equivalent `<meta>` element. It allows developers to restrict the sources from which resources such as JavaScript, CSS, images, files etc. are loaded. CSP is an effective defense in depth technique to mitigate the risk of vulnerabilities such as Cross Site Scripting (XSS) and Clickjacking.
```

> Content Security Policy can be implemented through one of the below methods,

Reviewer: ```suggestion
Content Security Policy can be implemented through one of two methods:
```

> Content Security Policy can be implemented through one of two methods:

Reviewer 2: ```suggestion
CSP can be implemented through one of two methods:
```

> - Content-Security-Policy Header: `Content-Security-Policy: policy`

Reviewer: ```suggestion
- `Content-Security-Policy` response header: `Content-Security-Policy: <policy>`
```

> - `Content-Security-Policy` response header: `Content-Security-Policy: <policy>`

Reviewer 2: ```suggestion
- CSP response header: `Content-Security-Policy: <policy>`
```

> - Meta element: `<meta http-equiv="Content-Security-Policy" content="policy">`

Reviewer: A note should be added here that not all directives can be applied via META, including but not necessarily limited to: `frame-ancestors`, `report-uri`, or `sandbox`

```suggestion
- Meta element: `<meta http-equiv="Content-Security-Policy" content="<policy>">`
```

> Content Security Policy supports directives which allow granular control to the flow of policies.

Reviewer 2: This whole part can be replaced with the CSP documentation, or the CSP cheat sheet, whichever you see as fitting to be added in the remediation, and not in the summary of the attack.
I believe this was discussed in another comment, just adding my 2 cents as well.

   ```HTTP
   script-src 'self' api.example.com; media-src media.example.com; img-src *
   ```

2. Enforce TLS for loading all resources.

   ```HTTP
   Content-Security-Policy: default-src https://secure.example.com
   ```

3. Enable violation reporting to URL specified in `report-uri`.

   ```HTTP
   Content-Security-Policy: default-src 'self'; report-uri http://reportcollector.example.com/collector.cgi
   ```

### Strict Policy

A strict policy is a policy which provides protection against classical stored, reflected, and some of the DOM XSS attacks and should be the optimal goal of any team trying to implement CSP.

Google went ahead and set up a guide to adopt a strict CSP based on nonces. Based on a presentation at [LocoMocoSec](https://speakerdeck.com/lweichselbaum/csp-a-successful-mess-between-hardening-and-mitigation?slide=55), the following two policies can be used to apply a strict policy:

Moderate Strict Policy:

```HTTP
script-src 'nonce-r4nd0m' 'strict-dynamic';
object-src 'none'; base-uri 'none';
```

Locked down Strict Policy:

```HTTP
script-src 'nonce-r4nd0m';
object-src 'none'; base-uri 'none';

> ```

Reviewer: I like this detail but it seems more developer focused than tester focused, I'm not sure testers need this level of background info. If they do perhaps we can just link to MDN or something?

This might be good for the cheat sheet project.

Author: Makes sense. I not very sure, how to cut it in a nice way though. 

But I think we can still callout strict policy, as it should be the ideal implementation of CSP.

Reviewer: Add the strict part to the remediation section?

Author: Yes. That's good. I will work on that

## Test Objectives

> Review the Content-Security-Policy header or meta element to identify misconfigurations.

Reviewer 2: ```suggestion
- Review the Content-Security-Policy header or meta element to identify misconfigurations.
```

## How to Test

> To test for misconfigurations in CSP policy look for below insecure configurations by examining `Content-Security-Policy` HTTP response header or CSP meta element in a proxy tool:

Reviewer: ```suggestion
To test for misconfigurations in content security policies look for below insecure configurations by examining `Content-Security-Policy` HTTP response header or CSP meta element in a proxy tool:
```

The P in the acronym is already policy. Saying CSP Policy is like saying SSN Number (Social Security Number Number) 😉

Author: :)

> To test for misconfigurations in content security policies look for below insecure configurations by examining `Content-Security-Policy` HTTP response header or CSP meta element in a proxy tool:

Reviewer 2: ```suggestion
To test for misconfigurations in CSPs look for below insecure configurations by examining `Content-Security-Policy` HTTP response header or CSP `meta` element in a proxy tool:
```

> - `unsafe-inline` directive enables inline scripts making the applications susceptible to Cross Site Scripting attacks.

Reviewer: ```suggestion
- `unsafe-inline` directive enables inline scripts or styles making the applications susceptible to Cross Site Scripting attacks.
```

> - `unsafe-inline` directive enables inline scripts or styles making the applications susceptible to Cross Site Scripting attacks.

Reviewer 2: ```suggestion
- `unsafe-inline` directive enables inline scripts or styles making the applications susceptible to XSS attacks.
```

> - `unsafe-eval` directive allows eval() to be used in the application.

Reviewer: ```suggestion
- `unsafe-eval` directive allows `eval()` to be used in the application.
```

> - Resources such as scripts can be allowed to be loaded from any origin by the use wildcard (*) source.
> - Framing can be enabled for all origins by the use of wildcard (*) source for `frame-ancestors` directive.
> - Business critical applications require to use a CSP strict policy.

Reviewer: ```suggestion
- Resources such as scripts can be allowed to be loaded from any origin by the use wildcard (`*`) source.
- Framing can be enabled for all origins by the use of wildcard (`*`) source for `frame-ancestors` directive.
- Business critical applications should require to use a strict policy.
```

> 

Reviewer 2: There is the bypass that occurs when URL whitelisting is used.
Let's say I whitelist a CDN to load my JS libraries, it could have a `jsonp` endpoint that can be used to bypass the CSP rules.

## Remediation

> Configure a strong CSP policy which reduces the attack surface of the application. Developers can verify the strength of content security policy using online tools such as [Google CSP Evaluator](https://csp-evaluator.withgoogle.com/).

Reviewer: ```suggestion
Configure a strong content security policy which reduces the attack surface of the application. Developers can verify the strength of content security policy using online tools such as [Google CSP Evaluator](https://csp-evaluator.withgoogle.com/).
```
